### PR TITLE
Add supporting infrastructure for first version of migo bio-tech

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
Infrastructure "Add supporting infrastructure for first version of migo bio-tech"

#### Purpose of change
Add functionality to json around items using effect_on_conditions.
Give example of use case for talker_item

#### Describe the solution
This adds a new migo bio tech item, which does not yet spawn anywhere.  The infrastructure needed for it to work has also been added.  
At present migo bio technology functions as follows: 
Upon usage it will choose a type which can be: "morale", "focus", "pain", or "fatigue".  It then gives you a bonus of that type.  
There is always a chance that a variable called migo interest goes up.  At present this does nothing but the plan is to have it eventually trigger mi go attacks.
Included infrastructure features:
- The ability to set a variable to a random value, 
- The ability to give an item conditional names based on the presence of a variable.
- Cleaned up morale function to take string times rather than int.
- Added support for defining in-line eocs for item usage. 

#### Describe alternatives you've considered
Not including the actual migo bio tech item.  It's not used yet and still needs better stats and details but it tests the new code and gives a good example of what can be done.

### CODE REVIEW

File: data/json/items/tool/science.json [hunk 1]
```diff
@@ -1342,5 +1342,100 @@
     "symbol": ";",
     "color": "light_gray",
     "use_action": [ "SEXTANT" ]
+  },
+  {
+    "id": "migo_bio_tech",
+    "type": "TOOL",
+    "name": { "str_sp": "Migo Bio Tech" },
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1349]:
```suggestion
    "name": { "str_sp": "Mi-go Bio Tech" },
```

File: data/json/items/tool/science.json [hunk 2]
```diff
@@ -1342,5 +1342,100 @@
...
+    "conditional_names": [
+      { "type": "VAR", "condition": "npctalk_var_DISPLAY_NAME_MORALE", "name": { "str_sp": "%s (happy)" } },
+      { "type": "VAR", "condition": "npctalk_var_DISPLAY_NAME_FOCUS", "name": { "str_sp": "%s (focus)" } },
+      { "type": "VAR", "condition": "npctalk_var_DISPLAY_NAME_PAIN", "name": { "str_sp": "%s (pain relief)" } },
+      { "type": "VAR", "condition": "npctalk_var_DISPLAY_NAME_FATIGUE", "name": { "str_sp": "%s (wake up)" } }
+    ],
+    "description": "A piece of migo bio technology.",
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1356]:
```suggestion
    "description": "A piece of mi-go biotechnology.",
```

File: data/json/items/tool/science.json [hunk 3]
```diff
@@ -1342,5 +1342,100 @@
...
+              "decay_start": "10 hours"
+            },
+            { "npc_add_var": "MORALE", "type": "DISPLAY", "context": "NAME", "value": "" }
+          ]
+        },
+        {
+          "id": "EOC_MIGO_BIO_TECH_FOCUS",
+          "condition": { "npc_has_var": "function", "type": "migo_bio_tech", "context": "bio_tech", "value": "focus" },
+          "effect": [
+            { "message": "You feel incredibily focused!" },
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1415]:
```suggestion
            { "message": "You feel incredibly focused!" },
```

File: data/json/items/tool/science.json [hunk 4]
```diff
@@ -1342,5 +1342,100 @@
...
+            { "message": "You feel your aches and pains fade away!" },
+            { "u_mod_pain": -30 },
+            { "npc_add_var": "PAIN", "type": "DISPLAY", "context": "NAME", "value": "" }
+          ]
+        },
+        {
+          "id": "EOC_MIGO_BIO_TECH_FATIGUE",
+          "condition": { "npc_has_var": "function", "type": "migo_bio_tech", "context": "bio_tech", "value": "fatigue" },
+          "effect": [
+            { "message": "You feel more awake than a moment ago!" },
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1433]:
```suggestion
            { "message": "You instantly feel more awake!" },
```

File: data/json/items/tool/science.json [hunk 5]
```diff
@@ -1342,5 +1342,100 @@
...
+          "effect": [
+            { "message": "You feel a little watched" },
+            { "u_adjust_var": "migo_interest", "type": "migo_interest", "context": "interest", "adjustment": 1 }
+          ]
+        },
+        {
+          "id": "EOC_MIGO_BIO_TECH_INTEREST",
+          "condition": { "not": { "one_in_chance": 100 } },
+          "effect": [
+            { "message": "You feel a lot watched" },
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1379]:
```suggestion
            { "message": "You have an intense feeling of being watched." },
```

File: data/json/items/tool/science.json [hunk 6]
```diff
@@ -1342,5 +1342,100 @@
...
+    "color": "light_red",
+    "use_action": {
+      "type": "effect_on_conditions",
+      "description": "Touch the opening.",
+      "effect_on_conditions": [
+        {
+          "id": "EOC_MIGO_BIO_TECH_INTEREST",
+          "condition": { "not": { "one_in_chance": 3 } },
+          "effect": [
+            { "message": "You feel a little watched" },
```
Comment (chaohedgehogs) [inline data/json/items/tool/science.json:1371]:
```suggestion
            { "message": "You have a slight feeling of being watched." },
```

File: data/json/morale_types.json [hunk 7]
```diff
@@ -388,5 +388,10 @@
     "id": "morale_nightmare",
     "type": "morale_type",
     "text": "Had a vivid nightmare"
+  },
+  {
+    "id": "morale_migo_bio_tech",
+    "type": "morale_type",
+    "text": "Used Migo Bio Tech"
```
Comment (chaohedgehogs) [inline data/json/morale_types.json:395]:
```suggestion
    "text": "Used mi-go biotech"
```

File: data/json/items/tool/science.json [hunk 8]
```diff
@@ -1356,5 +1356,100 @@
...
+    "color": "light_red",
+    "use_action": {
+      "type": "effect_on_conditions",
+      "description": "Touch the opening.",
+      "effect_on_conditions": [
+        {
+          "id": "EOC_MIGO_BIO_TECH_INTEREST",
+          "condition": { "not": { "one_in_chance": 3 } },
+          "effect": [
+            { "message": "You have a slight feeling of being watched." },
```
Comment (ZhilkinSerg) [inline data/json/items/tool/science.json:1385]: I am pretty sure it should be `u_message`, not `message`.

Comment (BrettDong) [reply data/json/items/tool/science.json:1385]: This is exactly what's causing #50590.

**Comment by hm001121606** [comment]
Can 'var' change more than 'name'? Such as' description 'or' use_action'😊

**Comment by Ramza13** [comment]
> Can 'var' change more than 'name'? Such as' description 'or' use_action'😊

No to either.  For description, that sounds like a good idea to add eventually, though I don't have a use case in mind right now so it will be pretty low on my to do list.  For use_action since with EOCs you can make what an item does change already I'd probably just add a new dialog function if needed.  If you have a use case in mind let me know and I'll see what I can do.

**Pull request merged**